xtractmime/_utils.py

```diff
@@ -0,0 +1,252 @@
+from struct import unpack
+from typing import Tuple, Union
+
+from xtractmime import _is_match_mime_pattern
+from xtractmime._patterns import (
+    IMAGE_PATTERNS,
+    AUDIO_VIDEO_PATTERNS,
+    FONT_PATTERNS,
+    ARCHIVE_PATTERNS,
+)
+
+sample_rates = [44100, 48000, 32000]
+mp25_rates = [
+    0,
+    8000,
+    16000,
+    24000,
+    32000,
+    40000,
+    48000,
+    56000,
+    64000,
+    80000,
+    96000,
+    112000,
+    128000,
+    144000,
+    160000,
+]
+mp3_rates = [
+    0,
+    32000,
+    40000,
+    48000,
+    56000,
+    64000,
+    80000,
+    96000,
+    112000,
+    128000,
+    160000,
+    192000,
+    224000,
+    256000,
+    320000,
+]
+
+
+def _is_mp4_signature(input_bytes: bytes) -> bool:
+    input_size = len(input_bytes)
+    if input_size < 12:
+        return False
+
+    box_size = unpack(">I", input_bytes[0:4])[0]
+
+    if (input_size < box_size) or (box_size % 4 != 0):
+        return False
+
+    if input_bytes[4:8] != b"ftyp":
+        return False
+
+    if input_bytes[8:11] == b"mp4":
+        return True
+
+    bytes_read = 16
+    while bytes_read < box_size:
+        if input_bytes[bytes_read : bytes_read + 3] == b"mp4":
+            return True
+        bytes_read += 4
+
+    return False
+
+
+def _parse_vint(input_bytes: bytes, input_size: int, index: int) -> int:
+    mask = 128
+    max_vint_size = 8
+    number_size = 1
+
+    while number_size < max_vint_size and number_size < input_size:
+        if input_bytes[index] & mask != 0:
+            break
+        mask = mask >> 1
+        number_size += 1
+
+    return number_size
+
+
+def _is_webm_signature(input_bytes: bytes) -> bool:
+    input_size = len(input_bytes)
+    if input_size < 4:
+        return False
+
+    if input_bytes[0:4] != b"\x1aE\xdf\xa3":
+        return False
+
+    index = 4
+
+    while index < input_size and index < 38:
+        if input_bytes[index : index + 2] == b"B\x82":
+            index += 2
+
+            if index >= input_size:
+                break
+
+            number_size = _parse_vint(input_bytes, input_size, index)
+            index += number_size
+
+            if index >= input_size - 4:
+                break
+
+            if input_bytes[index : index + 4] == b"webm":
+                return True
+        index += 1
+
+    return False
+
+
+def _match_mp3_header(input_bytes: bytes, input_size: int, index: int) -> bool:
+    if input_size < 4:
+        return False
+
+    if (
+        input_bytes[index : index + 1] != b"\xff"
+        or bytes([input_bytes[index + 1] & 224]) != b"\xe0"
```

Comment:
According to the standard:

```suggestion
        and bytes([input_bytes[index + 1] & 224]) != b"\xe0"
```

I strongly suggest adding a test for this; a test that would have failed with the current implementation, but passed after this fix.

Reply:
The algorithm mentioned in standards is wrong it should be `or` according to the `is_mp3` function below mentioned in mozzila implementation
```
static int is_mp3(const uint8_t* p, long length) {
 /* Do we have enough room to see a 4 byte header? */
      if (length < 4) return 0;
      /* Do we have a sync pattern? */
      if (p[0] == 0xff && (p[1] & 0xe0) == 0xe0) {
        /* Do we have any illegal field values? */
        if (((p[1] & 0x06) >> 1) == 0) return 0;  /* No layer 4 */
        if (((p[2] & 0xf0) >> 4) == 15) return 0; /* Bitrate can't be 1111 */
        if (((p[2] & 0x0c) >> 2) == 3) return 0;  /* Samplerate can't be 11 */
        /* Looks like a header. */
        if ((4 - ((p[1] & 0x06) >> 1)) != 3) return 0; /* Only want level 3 */
        return 1;
      }
  return 0;
}
```